Fix some chirality issues in SMARTS writing/parsing (Issues #2565 & #2568)

ricrogz opened the pull request:
This addresses #2565  & #2568.

It seems that my initial take on #2565 was right: `canonicalizeFragment` needs to know about `doIsomericSmiles` to be able to update the generated SMARTS chiralities according to the bond traversal order.

The problems I mentioned in my other PR (#2567, which I closed) came from two sides: the easy to fix one was that I was using patterns with explicit H atoms on SMILES that did not have them. As soon as I addded H to the SMILES, these went well.

The other problem was that, in the canonicalization, we compensate for implicit H on the leading chiral atom (if there is any), so we need to mirror that on the parser. But this is actually tricky, since we should only do that if we have a fully subscribed leading chiral atom (i.e. one that has three neigbors), since otherwise we would be describing a fragment, and we should take the SMARTS literally.

When I added my tests, I accidentally hit a tough case: "\[C@\](O)(F)Cl", which, when parsed as a SMILES, loses the chirality, but keeps it when parsed as SMARTS. I'm not sure how to handle that (maybe solve it some how in a different PR?), so I just commented it out in the test.

I'm also not too confident on my changes to the bison/flex parser. I'm sure there is a better way to achieve changes only on the first atom we see at all, but I'm not experienced with that parser, so I did what I could. Let me know if you can think of a better option and I'll update it.

greglandrum commented on Code/GraphMol/SmilesParse/smatest.cpp:
```diff
@@ -2671,11 +2671,81 @@ void testGithub2142() {
+       * This is a tough one: it won't keep chirality in SMILES.
+       * Let's skip it for now.
+       * R"([C@](O)(F)Cl)",
+       */
```
Just an info point: this loses chirality in the SMILES because the sanitization "knows" that a three-coordinate neutral C can't be chiral. So it gets removed.
Molecules built from SMARTS aren't sanitized on parsing, so this doesn't happen.

ricrogz replied:
Yeah, I imagined that. But my though was that, since we have 3 specified non-H neighbors and a chirality set, it would be reasonable to assume that the missing neighbor is an implicit H, and add it to the mol.

ricrogz replied:
I meant in SMILES, SMARTS should understand it as a fragment, and not add the implicit H.

greglandrum replied:
No, because by specifying the carbon with `[C@]`  you have explicitly said that it does *not* have any implicit Hs. That's how you specify radicals in SMILES.

so `[C@](O)(F)Cl` in SMILES corresponds to a C that has a radical electron.

ricrogz replied:
Oh, I see. I totally forgot about radicals!

greglandrum commented on Code/GraphMol/SmilesParse/smatest.cpp:
```diff
@@ -2671,11 +2671,81 @@ void testGithub2142() {
+    MolOps::addHs(*mol);  // We have explicit H in the queries!
+
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(SubstructMatch(*mol, *query, matches, true, true, true, false));
```
One of the habits I picked up from @bp-kelley is having local variables with the names of the optional arguments so that long streams of bools like we have here are actually understandable:
```suggestion
    bool uniquify=true,recursionPossible=true, useChirality=true;
    TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify, recursionPossible, useChirality));
```

greglandrum replied:
I intentionally left off the last argument (`useQueryQueryMatches`) since it was left in its default value.

ricrogz replied:
That makes sense. I'll update with it, although I don't like declaring more than one variable in the same line.

greglandrum replied:
me neither, I was just too lazy to do the suggestion across multiple lines in the github UI.

ricrogz replied:
heh, heh, no problem.

greglandrum commented on Code/GraphMol/SmilesParse/smarts.yy:
```diff
@@ -162,6 +162,13 @@ ATOM_OPEN_TOKEN bad_atom_def
+
+  // The first atom may require a chirality inversion if it has
+  // an implicit H. Mark it to be checked after the mol has been parsed
+  if(sz == 0 && $1->getChiralTag() != Atom::CHI_UNSPECIFIED) {
```
I think this is problematic for SMARTS.
The SMILES `[C@H](F)(Cl)C` may need an inversion because the H atom is actually there in the molecule and "participating in defining the chirality".
The SMARTS `[C@H](F)(Cl)C`, on the other hand, doesn't actually have an H affecting the chirality; it has a query that there's an H attached to the C.

ricrogz replied:
I see. I misinterpreted the H in the SMARTS, and this actually isn't an issue. I'll comment on #2568 and close it.

greglandrum commented:
Here's some behavior that happens both on master and your branch that I find odd : 
```
In [2]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('C[C@](Cl)(Br)F'),useChirality=True)                                   
Out[2]: True

In [3]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('C[C@@](Cl)(Br)F'),useChirality=True)                                  
Out[3]: False

In [4]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('[C@](Cl)(Br)F'),useChirality=True)                                    
Out[4]: False

In [5]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('[C@@](Cl)(Br)F'),useChirality=True)                                   
Out[5]: True
```
I'm pointing this out because I don't think I understand why it's happening and maybe could be looked at as part of this PR.
Or not. :-)

ricrogz commented:
> Here's some behavior that happens both on master and your branch that I find odd :
> 
> ```
> In [2]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('C[C@](Cl)(Br)F'),useChirality=True)                                   
> Out[2]: True
> 
> In [3]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('C[C@@](Cl)(Br)F'),useChirality=True)                                  
> Out[3]: False
> 
> In [4]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('[C@](Cl)(Br)F'),useChirality=True)                                    
> Out[4]: False
> 
> In [5]: Chem.MolFromSmiles('C[C@](Cl)(Br)F').HasSubstructMatch(Chem.MolFromSmarts('[C@@](Cl)(Br)F'),useChirality=True)                                   
> Out[5]: True
> ```
> 
> I'm pointing this out because I don't think I understand why it's happening and maybe could be looked at as part of this PR.
> Or not. :-)

I have been looking into this. The funny part is that there are tests enforcing this behavior:
https://github.com/rdkit/rdkit/blob/1daed4d16388037bd2e84c903134f8b82528cfb2/Code/GraphMol/Substruct/test1.cpp#L969 (this one, and the following ones till the end of the function).

I'm strying trying to decide what the right behavior should be in this situation (matching a SMARTS of a chiral atom which is only partially specified).

ricrogz commented:
Ok, in this update, I:
- reverted everything that was related to #2568.
- Updated the test for #2565 with your comments.
- Tracked back the issue you mentioned to MolMatchFinalCheckFunctor. I refactored it to help me understand what was going on there.
- Applied a fix for the issue, and also allowed chiral fragments with degree < 3 to be matched.
- Added a new test, and updated the old one that checked the wrong behavior. In the new test, I didn't check too much what should be matched or not, just went with my intuition (SMILES + SMARTS + Chirality = headache).

Please let me know what you think of this.

ricrogz commented:
I just noticed that patch breaks this case, which works on master:

```python
# coding: utf-8
from rdkit import Chem
smiles = r'[H][C@](O)(F)Cl'
mol = Chem.MolFromSmiles(smiles)

smarts = Chem.MolToSmarts(mol)
print(smarts)
print(Chem.MolToSmiles(mol))

query = Chem.MolFromSmarts(smarts)
assert mol.HasSubstructMatch(query, useChirality=True)
```

greglandrum commented:
That's because the `[H]` is being removed from the SMILES but not from the SMARTS:
```
In [4]: rdkit.__version__                                                                                              
Out[4]: '2019.03.3'

In [7]: ps = Chem.SmilesParserParams()                                                                                 

In [8]: ps.removeHs = False                                                                                            

In [9]: smiles = r'[H][C@](O)(F)Cl'                                                                                    

In [10]: m = Chem.MolFromSmiles(smiles,ps)                                                                             

In [11]: q = Chem.MolFromSmarts(smiles)                                                                                

In [13]: m.HasSubstructMatch(q,useChirality=True)                                                                      
Out[13]: True
```

greglandrum commented:
oops, sorry, didn't read that code snippet carefully enough.
Deleting that last comment.

ricrogz commented:
This second issue seems tricky to address, and also has some interaction with the one I initially tried to fix.

I have rewound to the initial fix, and added a little dirty trick to avoid problems with leading chiral atoms... seems to fix my snippet. I will see how the rest builds on this later on this afternoon/evening...

ricrogz commented:
This time everything seems to work fine.

I completely avoided writing chiral atoms to the first place of SMARTS strings. This should save us some trouble. And since SMARTS strings do not get canonicalized, it should be fine, I think.

greglandrum commented on Code/GraphMol/SmilesParse/SmartsWrite.cpp:
```diff
@@ -656,7 +656,26 @@ std::string FragmentSmartsConstruct(ROMol &mol, unsigned int atomIdx,
+        atom->getChiralTag() == Atom::CHI_TETRAHEDRAL_CW) {
+      auto nbrItr = mol.getAtomNeighbors(atom);
+      if (nbrItr.first != nbrItr.second) {
+        atomIdx = mol[*nbrItr.first]->getIdx();
```
This is a clever, lightly hacky, solution to the problem.
There's a very minor chance that this could still fail if the neighbor you pick also happens to be chiral. I'm not sure whether or not it's worth worrying about.

ricrogz replied:
Oh, you are right. I completely missed that the neighbors might also be chiral. I think we should consider that. I'll turn that into a loop to find the first non-chiral atom and start the SMARTS from there.

greglandrum commented on Code/GraphMol/SmilesParse/smatest.cpp:
```diff
@@ -2671,11 +2671,42 @@ void testGithub2142() {
+       R"(O=C1C[C@](Cl)1F)", R"(Br[C@@H](Cl)F)"});
+
+  for (const auto &smi : smiles) {
+    auto *mol = SmilesToMol(smi);
```
This line and line 2693 are going to leak core. It'd be better to use `std::unique_ptr` here

ricrogz replied:
Totally right. I was only thinking about chiralities and matches, and just missed the pointer there. I'll fix it.

greglandrum commented on Code/GraphMol/SmilesParse/smatest.cpp:
```diff
@@ -2671,11 +2671,42 @@ void testGithub2142() {
+
+    bool uniquify = true;
+    bool recursionPossible = true;
+    bool useChirality = true;
```
these will make @bp-kelley so happy. ;-)

greglandrum commented on Code/GraphMol/Substruct/SubstructMatch.cpp:
```diff
@@ -35,6 +35,20 @@
 namespace detail {
+
+namespace {
+bool hasChiralLabel(const Atom *at) {
```
```suggestion
bool hasChiralLabel(const Atom *at) {
  PRECONDITION(at, "no atom");
```

ricrogz replied:
Makes sense.

greglandrum commented on Code/GraphMol/Substruct/SubstructMatch.cpp:
```diff
@@ -35,6 +35,20 @@
+}
+
+//! Decide if an atom comes from a query that describes a mol fragment
+bool isFragmentAtom(const Atom *at) {
```
```suggestion
bool isFragmentAtom(const Atom *at) {
  PRECONDITION(at, "no atom");
```

greglandrum commented on Code/GraphMol/Substruct/SubstructMatch.cpp:
```diff
@@ -35,6 +35,20 @@
+         at->getChiralTag() == Atom::CHI_TETRAHEDRAL_CCW;
+}
+
+//! Decide if an atom comes from a query that describes a mol fragment
```
This is actually just testing whether or not the degree of the atom is <4 and it hasn't had the implicit valence value calculated.

greglandrum commented on Code/GraphMol/Substruct/SubstructMatch.cpp:
```diff
@@ -35,6 +35,20 @@
+//! Decide if an atom comes from a query that describes a mol fragment
+bool isFragmentAtom(const Atom *at) {
+  unsigned explicitDegree = at->getDegree() + at->getNumExplicitHs();
+  return explicitDegree < 4 && at->getImplicitValence() == -1;
```
Wouldn't this be just as valid if you used:
```suggestion
  return explicitDegree < 4 && at->getImplicitValence() <= 0;
```
A use case where this could happen: I could imagine manually constructing a query molecule by starting from a normal molecule and removing one atom from a chiral center and then re-sanitizing (making sure that chiral center has the noImplicit flag set). I think with your current logic this would end up matching anything when doing chiral searches. That doesn't seem right.

ricrogz replied:
Yeah. I need to reconsider all the "fragment" treatment in this PR. From the beginning of it, it has been the part I have been less confident on.

greglandrum commented on Code/GraphMol/Substruct/test1.cpp:
```diff
@@ -1589,8 +1589,212 @@ void testGithubIssue1489() {
+    {
+      const auto query = R"([C@@](Cl)Br)"_smarts;
+      std::vector<MatchVectType> matches;
+      TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
```
I think that the way the logic of tetrahedral chirality goes, matching less than three neighbors of a 4-coordinate chiral atom is *always* a match. There just isn't sufficient information in the query to tell you otherwise.

ricrogz replied:
Yeah, that makes sense, too. I think I took the order of the SMARTS string too literally here, thinking of matching a literal CW chiral atom in which the first two consecutive bonds to neighbors are Cl and Br, or the opposite: a CCW atom with the bonds in reverse order... anyway, as I said, I'm not 100% sure about the 'fragments' matching making sense.

greglandrum commented on Code/GraphMol/Substruct/test1.cpp:
```diff
@@ -1589,8 +1589,212 @@ void testGithubIssue1489() {
+    const auto mol = R"([H][C@](O)(F)Cl)"_smiles;
+    const auto query = R"([C@H](O)(F)Cl)"_smarts;
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(!SubstructMatch(*mol, *query, matches, uniquify,
```
I believe that this is incorrect. 
Since the SMARTS `[C@](O)(F)Cl` matches, `[C@H](O)(F)Cl` definitely should do so as well.

ricrogz replied:
Right. I don't know where my thoughts were on that one. I probably got confused with the matching between the neighboring H and the explicit H... somehow.

ricrogz replied:
I'm reviewing this, and thinking: how is this case different from the one in one of your previous comments:

> The SMILES `[C@H](F)(Cl)C` may need an inversion because the H atom is actually there in the molecule and "participating in defining the chirality".
> The SMARTS `[C@H](F)(Cl)C`, on the other hand, doesn't actually have an H affecting the chirality; it has a query that there's an H attached to the C.

This is almost exactly the same case, so according to that, this test was correct, and neither SMARTS [C@](O)(F)Cl nor [C@H](O)(F)Cl should match SMILES [C@H](O)(F)Cl, as the SMILES will have its chirality reversed??

ricrogz commented:
- Updated with current master.
- Added a patch to consider #2594 in the chiral atom handling in the canonicalization, and also to update the tests associated with that PR.
- Fix leaking pointers in tests.
- Fix not using chiral atoms as lead of SMARTS strings (we now use the first non-chiral atom we can find).
- Remove faulty "fragments" matching logic.

greglandrum commented:
@ricrogz Sorry it took so long, but I think this is ready to merge (once the conflicts are resolved)
That seem right to you too?

ricrogz commented:
> @ricrogz Sorry it took so long, but I think this is ready to merge (once the conflicts are resolved)
> That seem right to you too?

Don't worry. I prefer it to take longer and be right, that merging it right in if we are not sure it is right.

Speaking of which, it might be because this has been a long, long week for me, but right now I can't make sense of some of the checks in Code/GraphMol/Substruct/test1.cpp... I guess I'll go get some rest, and maybe tomorrow, with fresh eyes, this makes sense again.

ricrogz commented:
> > @ricrogz Sorry it took so long, but I think this is ready to merge (once the conflicts are resolved)
> > That seem right to you too?
> 
> Don't worry. I prefer it to take longer and be right, that merging it right in if we are not sure it is right.
> 
> Speaking of which, it might be because this has been a long, long week for me, but right now I can't make sense of some of the checks in Code/GraphMol/Substruct/test1.cpp... I guess I'll go get some rest, and maybe tomorrow, with fresh eyes, this makes sense again.

Ok, now I saw why this didn't make sense yesterday: I had not updated with the changes regarding the question about matching `[C@H](O)(F)Cl` against `[H][C@](O)(F)Cl)`. I rechecked them, and they make sense.

@greglandrum, if the last commit I just added makes sense to you too, then I am all ok with merging.

greglandrum commented:
Hmm, the testing code made sense to me before; now it confuses me. Given that I/we keep getting mixed up about this, I think it’s worth putting together a little documentation, perhaps including pictures describing what we think we think should happen in each situation.  We can use this here to make sure we think we have it right and then to refer to later next time there’s a bug. :-)

greglandrum commented:
As a quick starting point:

 mol SMILES | query SMARTS | should match? 
——————|———————|————————
`C[C@](F)(Cl)O` | `C[C@](F)(Cl)O` | yes
`C[C@](F)(Cl)O` | `[C@](F)(Cl)O` | yes
`C[C@](F)(Cl)O` | `C[C@@](F)(Cl)O` | no
`C[C@](F)(Cl)O` | `[C@@](F)(Cl)O` | no
`C[C@](F)(Cl)O` | `[C@](Cl)O` | yes
`C[C@](F)(Cl)O` | `[C@@](Cl)O` | yes
`C[C@](F)(Cl)O` | `C[C@@](Cl)(F)O` | yes
`C[C@](F)(Cl)O` | `[C@@](Cl)(F)O` | yes


(More coming)

ricrogz commented:
Not really...

My two cents:

mol SMILES | query SMARTS | should match? | notes
---------- | ------------ | ------------- | -----
`C[C@](F)(Cl)O` | `C[C@](F)(Cl)O` | yes |
`C[C@](F)(Cl)O` | `[C@](F)(Cl)O` | yes | disagree (1)
`C[C@](F)(Cl)O` | `C[C@@](F)(Cl)O` | no |
`C[C@](F)(Cl)O` | `[C@@](F)(Cl)O` | no | disagree (2)
`C[C@](F)(Cl)O` | `[C@](Cl)O` | yes |
`C[C@](F)(Cl)O` | `[C@@](Cl)O` | yes |
`C[C@](F)(Cl)O` | `C[C@@](Cl)(F)O` | yes |
`C[C@](F)(Cl)O` | `[C@@](Cl)(F)O` | yes | disagree (3)
`[C@](C)(F)(Cl)O` | `C[C@](F)(Cl)O` | yes | `[C@](C)(F)(Cl)O` == `C[C@](F)(Cl)O`
`[C@](C)(F)(Cl)O` | `[C@](C)(F)(Cl)O` | yes | `[C@](C)(F)(Cl)O` == `C[C@](F)(Cl)O`

(3) is essentially the same as (1), and (2) is opposite to (1), so behavior should also be the opposite.

I see case (1) as follows:

According to the explanations in "3.3.3. Configuration Around Tetrahedral Centers" in  https://www.daylight.com/dayhtml/doc/theory/theory.smiles.html, the reference mol SMILES `C[C@](F)(Cl)O` should look something like this:

<img width="171" alt="case" src="https://user-images.githubusercontent.com/7498185/63640214-967bb200-c66b-11e9-81dd-756e1d7ba396.png">

While the query SMARTS `[C@](F)(Cl)O` should be something like the following (the asterisk symbolizes the unspecified fourth neighbor to the chiral atom):

<img width="171" alt="case2" src="https://user-images.githubusercontent.com/7498185/63640217-9a0f3900-c66b-11e9-804b-20c8f413e5bc.png">

To better compare, if we switch to looking down the bond between the unspecified neighbor and the chiral center (the rotation direction is irrelevant, but I included it anyway):

<img width="120" alt="case3" src="https://user-images.githubusercontent.com/7498185/63640266-0d18af80-c66c-11e9-853a-206d8af749a9.png">

Which is different from the reference mol (can be seen even better if rotated 180º).

Did I get something wrong?

greglandrum commented:
I think that the question is where we should think about the fourth attachment point going in that SMARTS.

I believe that we should think about it so that:`[C@](F)(Cl)O` matches the same structures that `*[C@](F)(Cl)O` matches (i.e. the fourth attachment point goes at the beginning)

ricrogz commented:
> I think that the question is where we should think about the fourth attachment point going in that SMARTS.
> 
> I believe that we should think about it so that:`[C@](F)(Cl)O` matches the same structures that `*[C@](F)(Cl)O` matches (i.e. the fourth attachment point goes at the beginning)
> 
> Note that this is just my gut feeling and is an answer that, I think, makes sense.
> The question of what is "right" is more difficult, and hard to answer

That is definitely the other option. The downside I see to it is that is the opposite of the behavior that the RDKit has right now:

https://github.com/rdkit/rdkit/blob/574caeba572e98b6863713c657739064e3dfd3e4/Code/GraphMol/Substruct/test1.cpp#L970-L981

The "pro" I see is that it seems more intuitive that `[C@](F)(Cl)O` should match `C[C@](F)(Cl)O` because of the great similarity of both strings.

The fun fact is that, then, I see no reason for giving SMILES `[H][C@](F)(Cl)O` == `[C@H](F)(Cl)O` a different treatment to `C[C@](F)(Cl)O`. So, SMARTS `[C@](F)(Cl)O` should also match that those, and also `[C@H](F)(Cl)O` in case of the H ones. And just like that, we circled all the way back to #2568, and we would have that using the same string for SMILES and SMARTS should yield a match between them.

greglandrum commented:
To wrap up the decision @ricrogz and I made here:
We will merge this PR, which doesn't change existing substructure matching behavior, as is and then re-examine how substructure matching between a 3-coordinate query and a 4-coordinate molecule should work for the next release. 

The reason this is worth mentioning is that we've learned that what the RDKit currently does is inconsistent with what it seems the Daylight substructure matching code did. We should decide what we want to do about that. Perhaps it's changing the current behavior, perhaps it's just documenting things better.
